server/src/main/java/io/deephaven/server/arrow/BarrageSubscriptionRequestHandler.java

```diff
@@ -0,0 +1,210 @@
+//
+package io.deephaven.server.arrow;
+
+import com.google.auto.service.AutoService;
+import com.google.rpc.Code;
+import io.deephaven.barrage.flatbuf.BarrageMessageType;
+import io.deephaven.barrage.flatbuf.BarrageSubscriptionRequest;
+import io.deephaven.engine.liveness.LivenessReferent;
+import io.deephaven.engine.table.impl.perf.QueryPerformanceNugget;
+import io.deephaven.engine.table.impl.perf.QueryPerformanceRecorder;
+import io.deephaven.extensions.barrage.BarrageMessageWriter;
+import io.deephaven.extensions.barrage.BarrageSubscriptionOptions;
+import io.deephaven.extensions.barrage.util.BarrageProtoUtil;
+import io.deephaven.extensions.barrage.util.GrpcUtil;
+import io.deephaven.proto.util.Exceptions;
+import io.deephaven.proto.util.ExportTicketHelper;
+import io.deephaven.server.session.SessionState;
+import io.deephaven.server.session.TicketRouter;
+import io.deephaven.util.SafeCloseable;
+import io.grpc.stub.StreamObserver;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * Handler for BarrageSubscriptionRequest over DoExchange.
+ */
+class BarrageSubscriptionRequestHandler implements ArrowFlightUtil.DoExchangeMarshaller.Handler {
+
+    private final ArrowFlightUtil.DoExchangeMarshaller marshaller;
+    private final TicketRouter ticketRouter;
+    private final SessionState session;
+    private final StreamObserver<BarrageMessageWriter.MessageView> listener;
+
+    /**
+     * The object representing the subscription.
+     */
+    private ExchangeMarshaller.Subscription subscriptionObject;
+
+    private Queue<BarrageSubscriptionRequest> preExportSubscriptions;
+    private SessionState.ExportObject<?> onExportResolvedContinuation;
+
+    public BarrageSubscriptionRequestHandler(final ArrowFlightUtil.DoExchangeMarshaller marshaller,
+            final TicketRouter ticketRouter,
+            final SessionState session,
+            final StreamObserver<BarrageMessageWriter.MessageView> listener) {
+        this.marshaller = marshaller;
+        this.ticketRouter = ticketRouter;
+        this.session = session;
+        this.listener = listener;
+    }
+
+    @Override
+    public void handleMessage(@NotNull final BarrageProtoUtil.MessageInfo message) {
+        // verify this is the correct type of message for this handler
+        if (message.app_metadata.msgType() != BarrageMessageType.BarrageSubscriptionRequest) {
+            throw Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT,
+                    "Request type cannot be changed after initialization, expected BarrageSubscriptionRequest metadata");
+        }
+
+        if (message.app_metadata.msgPayloadVector() == null) {
+            throw Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT, "Subscription request not supplied");
+        }
+
+        // ensure synchronization with parent class functions
+        synchronized (marshaller) {
```

Comment:
`#onExportResolved` is invoked from another thread; do we also need to synchronize on the marshaller for any of those marshaller member calls?

Reply:
I believe that I kept the synchronization unchanged from the original here:

https://github.com/deephaven/deephaven-core/pull/6736/files#diff-a9c0e91bd26dc6e2ab52324f8922229e40b2307c78e59b9b0ddc3ee861ee86d8L725

Which is excluded against changes to onExportResolvedContinuation in handleMessage.

Reply:
Ah, okay. This synchronization looks intended to prevent simultaneous invocations to `handleMessage`. Specifically, we're avoiding races on subscription requests (we want to know which is newest):
```java
                    // have we already created the queue?
                    if (preExportSubscriptions != null) {
                        preExportSubscriptions.add(subscriptionRequest);
                        return;
                    }
```

Interestingly enough all callers are already synchronized in the marshaller before invoking `handleMessage`. Maybe we should remove the superfluous synchronization on the parent on the inner impls? As is, we might keep repeating an unnecessary pattern.

Reply:
I removed the inner sync on the marshaller, which I agree was already done before handleMessage, does this make sense to you now?